docs/academy/exercise5.md

```diff
@@ -1,20 +1,398 @@
+        (void) user_ptr;
+
+        assert(update_initialized);
+        update_initialized = false;
+
+        int res = flash_aligned_writer_flush(&writer);
+        if (res) {
+            avs_log(fw_update, ERROR,
+                    "Failed to finish download: flash aligned writer flush failed, "
+                    "result: %d",
+                    res);
+            return -1;
+        }
+
+        return 0;
+    }
+    ```
+
+
+- `reset` is called whenever there is an error during firmware download, or if the Server decides to not pursue firmware update with downloaded firmware (e.g. because it was notified that firmware verification failed).
+
+    <p style="text-align: center;">firmware_update.c</p>
+    ``` c
+    static void fw_reset(void *user_ptr) {
+        (void) user_ptr;
+
+        update_initialized = false;
+    }
+
+    static void fw_update_reboot(avs_sched_t *sched, const void *data) {
+        (void) sched;
+        (void) data;
+
+        avs_log(fw_update, INFO, "Rebooting.....");
+        pfb_perform_update();
+    }
+    ```
+
+
+- `perform_upgrade` is called whenever the download is finished, the firmware is successfully verified on the Client and the Server decides to upgrade the device.
+
+    <p style="text-align: center;">firmware_update.c</p>
+    ``` c
+    static int fw_perform_upgrade(void *anjay) {
+        pfb_mark_download_slot_as_valid();
+        avs_log(fw_update, INFO,
+                "The firmware will be updated at the next device reset");
+
+        return AVS_SCHED_DELAYED(anjay_get_scheduler(anjay), NULL,
+                                avs_time_duration_from_scalar(1, AVS_TIME_S),
+                                fw_update_reboot, NULL, 0);
+    }
+
+    ```
+
+To install the module, we are going to use the *fw_update_install()* function which is called in the main.c file:
+
+<p style="text-align: center;">firmware_update.c</p>
+``` c
+static const anjay_fw_update_handlers_t handlers = {
+    .stream_open = fw_stream_open,
+    .stream_write = fw_stream_write,
+    .stream_finish = fw_stream_finish,
+    .reset = fw_reset,
+    .perform_upgrade = fw_perform_upgrade
+};
+int fw_update_install(anjay_t *anjay) {
+    anjay_fw_update_initial_state_t state = { 0 };
+
+    if (pfb_is_after_firmware_update()) {
+        state.result = ANJAY_FW_UPDATE_INITIAL_SUCCESS;
+        avs_log(fw_update, INFO, "Running on a new firmware");
+    }
+    if (pfb_firmware_sha256_check(downloaded_bytes)) {
+        avs_log(fw_update, ERROR, "SHA256 check failed");
+        return -1;
+    }
+
+    return anjay_fw_update_install(anjay, &handlers, anjay, &state);
+}
+```
+
+## Flash alignment
+
+Flash APIs require that the length of data to write will be a multiple of 256 bytes, so we need to enforce that by additional buffering. For this in the firmware_update directory there are two more files:
+
+- flash_aligned_writer.c
+
+<p style="text-align: center;">flash_aligned_writer.c</p>
+``` c
+    #include <assert.h>
```

Comment:
You've indented the code, but not the whole section. It renders badly. Applies for other code snippets.